File: Python/credit.py
def main():
    number = input("Number: ").strip()

    if not is_valid_by_luhn(number):
        print("INVALID")

    elif len(number) == 15 and number[:2] in ("34", "37"):
        print("AMEX")

    elif len(number) == 16 and number[:2] in ("51", "52", "53", "54", "55"):
        print("MASTERCARD")

    elif len(number) in (13, 16) and number[0] == "4":
        print("VISA")

    else:
        print("INVALID")


# Hans Peter Luhn algorithm that checks card validity
def is_valid_by_luhn(number):
    sum = get_sum(int(number))
    return sum % 10 == 0


def get_sum(number):
    index = sum = 0

    while number > 0:
        digit = number % 10  # get last digit
        number //= 10  # truncate last digit from number

        if digit == 0:
            index += 1
            continue

        if index % 2 != 0:
            digit *= 2  # multiply digits from odd positions

            if digit >= 10:
                digit = digit % 10 + digit // 10

        sum += digit  # add all digits from even positions and multiplied digits from odd postitions
        index += 1

    return sum


if __name__ == "__main__":
    main()

File: C/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int main(void)
{
    long number = get_long("Number: ");

    long amex_length_15 = pow(10, 13);
    long mc_length_16 = pow(10, 14);
    long visa_length_13 = pow(10, 12);
    long visa_length_16 = pow(10, 15);

    //Hans Peter Luhn algorithm that checks card validity
    int index = 0, sum = 0;
    for (long cut_num = number; cut_num > 0; cut_num /= 10, index++)
    {
        int digit = cut_num % 10; //getting every digit
        if (index % 2 != 0)
        {
            int multiply = digit * 2; //multiplying every other digit starting from the end
            if (multiply >= 10)
            {
                multiply = multiply % 10 + multiply / 10;
            }

            sum += multiply; //adding products digits
        }
        else
        {
            sum += digit; //adding the remaining digits
        }
    }

    if (sum % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }

    //Checking validity and identifying card type based on its first numbers and length
    if
    (
        ((number >= 34 * amex_length_15) && (number < 35 * amex_length_15))
        ||
        ((number >= 37 * amex_length_15) && (number < 38 * amex_length_15))
    )
    {
        printf("AMEX\n");
    }
    else if ((number >= 51 * mc_length_16) && (number < 56 * mc_length_16))
    {
        printf("MASTERCARD\n");
    }
    else if
    (
        ((number >= 4 * visa_length_13) && (number < 5 * visa_length_13))
        ||
        ((number >= 4 * visa_length_16) && (number < 5 * visa_length_16))
    )
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}